File: plugins/tools/r_heatmap/runner.R
#!/usr/bin/env Rscript
# R heatmap visualization workspace tool

args <- commandArgs(trailingOnly = TRUE)

tool_kind <- "heatmap"
input_path <- NULL
output_path <- NULL
check_deps <- FALSE

i <- 1
while (i <= length(args)) {
  if (args[i] == "--tool-kind" && i + 1 <= length(args)) {
    tool_kind <- args[i + 1]
    i <- i + 2
  } else if (args[i] == "--input" && i + 1 <= length(args)) {
    input_path <- args[i + 1]
    i <- i + 2
  } else if (args[i] == "--output" && i + 1 <= length(args)) {
    output_path <- args[i + 1]
    i <- i + 2
  } else if (args[i] == "--check-deps") {
    check_deps <- TRUE
    i <- i + 1
  } else {
    i <- i + 1
  }
}

required <- c("ggplot2", "readr", "pheatmap")
missing <- required[!vapply(required, requireNamespace, quietly = TRUE, FUN.VALUE = logical(1))]

if (check_deps) {
  if (length(missing) > 0) {
    cat(jsonlite::toJSON(list(status = "missing", packages = missing), auto_unbox = TRUE), "\n")
    quit(status = 2)
  }
  cat(jsonlite::toJSON(list(status = "ok", packages = required), auto_unbox = TRUE), "\n")
  quit(status = 0)
}

if (length(missing) > 0) {
  cat("Missing optional R dependencies:", paste(missing, collapse = ", "), "\n")
  cat("Install them in your workspace R library before running this tool.\n")
  quit(status = 2)
}

suppressPackageStartupMessages({
  library(ggplot2)
  library(readr)
  library(pheatmap)
})

if (!is.null(input_path) && file.exists(input_path)) {
  df <- read_delim(input_path, delim = ifelse(grepl("\\.tsv$", input_path), "\t", ","), show_col_types = FALSE)
} else {
  set.seed(42)
  df <- data.frame(
    gene_a = c(1.2, 3.4, 2.1, 4.5, 0.8),
    gene_b = c(2.3, 1.1, 4.2, 0.9, 3.6),
    gene_c = c(3.1, 2.8, 1.5, 3.2, 2.0),
    gene_d = c(0.5, 4.1, 2.9, 1.3, 3.8),
    row.names = paste0("sample_", 1:5)
  )
}

numeric_cols <- vapply(df, is.numeric, logical(1))
mat <- as.matrix(df[, numeric_cols, drop = FALSE])

if (!is.null(output_path)) {
  png(output_path, width = 800, height = 600, res = 150)
  pheatmap(mat, main = "Heatmap", cluster_rows = TRUE, cluster_cols = TRUE)
  dev.off()
  cat(jsonlite::toJSON(list(status = "ok", output = output_path, shape = dim(mat)), auto_unbox = TRUE), "\n")
} else {
  cat(jsonlite::toJSON(list(
    status = "ok",
    shape = dim(mat),
    columns = colnames(mat),
    message = "Heatmap generated. Use --output to save as PNG."
  ), auto_unbox = TRUE), "\n")
}

File: plugins/tools/heatmap/runner.py
#!/usr/bin/env python3
"""Python heatmap visualization workspace tool."""

from __future__ import annotations

import argparse
import importlib.util
import json
import sys
from pathlib import Path


def missing(packages: list[str]) -> list[str]:
    return [pkg for pkg in packages if importlib.util.find_spec(pkg) is None]


def main() -> int:
    parser = argparse.ArgumentParser(description="SuperMedicine Python heatmap tool")
    parser.add_argument("--input", default=None, help="CSV/TSV data matrix path")
    parser.add_argument("--output", default=None, help="Output PNG image path")
    parser.add_argument("--tool-kind", default="heatmap")
    parser.add_argument("--check-deps", action="store_true")
    args = parser.parse_args()

    required = ["pandas", "matplotlib", "seaborn"]
    if args.check_deps:
        unavailable = missing(required)
        if unavailable:
            print(json.dumps({"status": "missing", "packages": unavailable}))
            return 2
        print(json.dumps({"status": "ok", "packages": required}))
        return 0

    unavailable = missing(required)
    if unavailable:
        print(
            f"Missing optional Python dependencies: {', '.join(unavailable)}",
            file=sys.stderr,
        )
        print("Install them before running this tool.", file=sys.stderr)
        return 2

    try:
        import matplotlib
        matplotlib.use("Agg")
        import matplotlib.pyplot as plt
        import pandas as pd
        import seaborn as sns  # type: ignore[import-untyped]

        if args.input:
            input_path = Path(args.input)
            if not input_path.is_file():
                raise FileNotFoundError(f"input file not found: {args.input}")
            sep = "\t" if input_path.suffix.lower() in (".tsv", ".tab") else ","
            df = pd.read_csv(input_path, sep=sep)
        else:
            df = pd.DataFrame(
                {
                    "gene_a": [1.2, 3.4, 2.1, 4.5, 0.8],
                    "gene_b": [2.3, 1.1, 4.2, 0.9, 3.6],
                    "gene_c": [3.1, 2.8, 1.5, 3.2, 2.0],
                    "gene_d": [0.5, 4.1, 2.9, 1.3, 3.8],
                },
                index=["sample_1", "sample_2", "sample_3", "sample_4", "sample_5"],
            )

        numeric_df = df.select_dtypes(include="number")
        if numeric_df.empty:
            raise ValueError("no numeric columns found for heatmap")

        fig, ax = plt.subplots(figsize=(8, 6))
        sns.heatmap(numeric_df, annot=True, fmt=".2f", cmap="YlOrRd", ax=ax)
        ax.set_title("Heatmap")

        if args.output:
            output_path = Path(args.output)
            output_path.parent.mkdir(parents=True, exist_ok=True)
            fig.savefig(output_path, dpi=150, bbox_inches="tight")
            result = {"status": "ok", "output": str(output_path), "shape": list(numeric_df.shape)}
        else:
            result = {
                "status": "ok",
                "shape": list(numeric_df.shape),
                "columns": list(numeric_df.columns),
                "message": "Heatmap generated. Use --output to save as PNG.",
            }
        plt.close(fig)
        print(json.dumps(result, ensure_ascii=False, indent=2))
        return 0
    except Exception as exc:
        print(json.dumps({"status": "error", "error": str(exc)}), file=sys.stderr)
        return 2


if __name__ == "__main__":
    raise SystemExit(main())
